pkg/connection/manager.go

```diff
@@ -0,0 +1,154 @@
+	return func(m *DetailsManager) {
+		m.log = l
+	}
+}
+
+// WithStoreBuilder configures the StoreBuilder to use.
+func WithStoreBuilder(sb StoreBuilderFn) DetailsManagerOption {
+	return func(m *DetailsManager) {
+		m.storeBuilder = sb
+	}
+}
+
+// DetailsManager is a connection details manager that satisfies the required
+// interfaces to work with connection details by managing interaction with
+// different store implementations.
+type DetailsManager struct {
+	client       client.Client
+	newConfig    func() StoreConfig
+	storeBuilder StoreBuilderFn
+
+	log logging.Logger
+}
+
+// NewDetailsManager returns a new connection DetailsManager.
+func NewDetailsManager(c client.Client, of schema.GroupVersionKind, o ...DetailsManagerOption) *DetailsManager {
+	nc := func() StoreConfig {
+		return resource.MustCreateObject(of, c.Scheme()).(StoreConfig)
+	}
+
+	m := &DetailsManager{
+		client:       c,
+		newConfig:    nc,
+		storeBuilder: RuntimeStoreBuilder,
+
+		log: logging.NewNopLogger(),
+	}
+
+	for _, mo := range o {
+		mo(m)
+	}
+
+	return m
+}
+
+// PublishConnection publishes the supplied ConnectionDetails to a secret on
+// the configured connection Store.
+// TODO(turkenh): Refactor this method once existing interface methods refactored.
+func (m *DetailsManager) PublishConnection(ctx context.Context, mg resource.Managed, c managed.ConnectionDetails) error {
+	return m.publishConnection(ctx, mg.(SecretOwner), store.KeyValues(c))
+}
+
+// UnpublishConnection deletes connection details secret from the configured
+// connection Store.
+// TODO(turkenh): Refactor this method once existing interface methods refactored.
+func (m *DetailsManager) UnpublishConnection(ctx context.Context, mg resource.Managed, c managed.ConnectionDetails) error {
+	return m.unpublishConnection(ctx, mg.(SecretOwner), store.KeyValues(c))
+}
+
+func (m *DetailsManager) connectStore(ctx context.Context, p *v1.PublishConnectionDetailsTo) (Store, error) {
+	sc := m.newConfig()
+	if err := m.client.Get(ctx, types.NamespacedName{Name: p.SecretStoreConfigRef.Name}, sc); err != nil {
+		return nil, errors.Wrap(err, errGetStoreConfig)
+	}
+
+	return m.storeBuilder(ctx, m.client, sc.GetStoreConfig())
+}
+
+func (m *DetailsManager) publishConnection(ctx context.Context, so SecretOwner, kv store.KeyValues) error {
+	// This resource does not want to expose a connection secret.
+	p := so.GetPublishConnectionDetailsTo()
+	if p == nil {
+		return nil
+	}
+
+	ss, err := m.connectStore(ctx, p)
+	if err != nil {
+		return errors.Wrap(err, errConnectStore)
+	}
+
+	return errors.Wrap(ss.WriteKeyValues(ctx, store.Secret{
+		Name:     p.Name,
+		Scope:    so.GetNamespace(),
+		Metadata: p.Metadata.Raw,
+	}, kv), errWriteStore)
+}
+
+func (m *DetailsManager) unpublishConnection(ctx context.Context, so SecretOwner, kv store.KeyValues) error {
+	// This resource didn't expose a connection secret.
+	p := so.GetPublishConnectionDetailsTo()
+	if p == nil {
+		return nil
+	}
```

Comment:
I suppose it's possible that a resource could have published connection details, but then later unset its `spec.publishConnectionDetailsTo` field. Previously owner ref based GC would have removed the connection secret anyhow, but in this implementation we'd presumably leak them. Not sure this is really a problem though - it's  quite an edge case.